File: speech_to_text/csharp/transcribe_file/Local/Local.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

string apiBase = "https://api.soniox.com";
string fileToTranscribe = "coffee_shop.mp3";

// Retrieve the API key from the environment variable
string? apiKey = Environment.GetEnvironmentVariable("SONIOX_API_KEY");
if (string.IsNullOrEmpty(apiKey))
{
    throw new InvalidOperationException("SONIOX_API_KEY is not set.");
}

// Create a requests session and set the Authorization header
var client = new HttpClient();
client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

// 1. Upload file
Console.WriteLine("Starting file upload...");
using var fileStream = new FileStream(fileToTranscribe, FileMode.Open, FileAccess.Read);
var fileContent = new MultipartFormDataContent();
fileContent.Add(new StreamContent(fileStream), "file", Path.GetFileName(fileToTranscribe));

var response = await client.PostAsync($"{apiBase}/v1/files", fileContent);
response.EnsureSuccessStatusCode();
var responseJson = await response.Content.ReadAsStringAsync();
var file = JsonNode.Parse(responseJson);

// Ensure file is parsed correctly
var fileId = file?["id"]?.ToString();
if (string.IsNullOrEmpty(fileId))
{
    throw new InvalidOperationException("File ID is missing after upload.");
}

// 2. Start a new transcription session
Console.WriteLine("Starting transcription...");
var request = new StringContent(
    JsonSerializer.Serialize(new { file_id = fileId, model = "stt-async-preview" }),
    System.Text.Encoding.UTF8,
    "application/json"
);
response = await client.PostAsync($"{apiBase}/v1/transcriptions", request);
response.EnsureSuccessStatusCode();
responseJson = await response.Content.ReadAsStringAsync();

var transcription = JsonNode.Parse(responseJson);
var transcriptionId = transcription?["id"]?.ToString();
if (string.IsNullOrEmpty(transcriptionId))
{
    throw new InvalidOperationException("Transcription ID is missing.");
}

Console.WriteLine($"Transcription started with ID: {transcriptionId}");

// 3. Poll the transcription endpoint until the status is 'completed'
while (true)
{
    response = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}");
    response.EnsureSuccessStatusCode();
    responseJson = await response.Content.ReadAsStringAsync();
    transcription = JsonNode.Parse(responseJson);

    var status = transcription?["status"]?.ToString();
    if (string.IsNullOrEmpty(status))
    {
        throw new InvalidOperationException("Status is missing in transcription response.");
    }

    if (status == "error")
    {
        var errorMessage = transcription?["error_message"]?.ToString() ?? "Unknown error";
        throw new Exception($"Transcription error: {errorMessage}");
    }
    else if (status == "completed")
    {
        break;
    }

    await Task.Delay(1000);
}

// 4. Retrieve the final transcript
response = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}/transcript");
response.EnsureSuccessStatusCode();
responseJson = await response.Content.ReadAsStringAsync();

var transcript = JsonNode.Parse(responseJson);
var transcriptText = transcript?["text"]?.ToString();
if (string.IsNullOrEmpty(transcriptText))
{
    throw new InvalidOperationException("Transcript text is missing.");
}

Console.WriteLine("Transcript:");
Console.WriteLine(transcriptText);

File: speech_to_text/php/transcribe_file/local/local.php
<?php

$api_base = "https://api.soniox.com";
$api_key = getenv("SONIOX_API_KEY");

function create_curl_session($url, $headers = [], $post_fields = null, $is_post = false, $is_json = false) {
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_URL, $url);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
    if ($is_post) {
        curl_setopt($curl, CURLOPT_POST, true);
        if ($is_json) {
            curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($post_fields));
        } else {
            curl_setopt($curl, CURLOPT_POSTFIELDS, $post_fields);
        }
    }
    return $curl;
}

$auth_header = ["Authorization: Bearer $api_key"];

// 1. Upload a file
echo "Starting file upload...\n";
$session = create_curl_session(
    "$api_base/v1/files",
    $auth_header,
    ["file" => new CURLFile("coffee_shop.mp3")],
    true
);
$res = curl_exec($session);
$http_status = curl_getinfo($session, CURLINFO_HTTP_CODE);
curl_close($session);

if ($http_status !== 201) {
    die("File upload failed: $res");
}

$file = json_decode($res, true);

// 2. Start a new transcription session by sending the audio URL to the API
echo "Starting transcription...\n";
$session = create_curl_session(
    "$api_base/v1/transcriptions",
    array_merge($auth_header, ["Content-Type: application/json"]),
    [
        "file_id" => $file["id"],
        "model" => "stt-async-preview"
    ],
    true,
    true
);

$res = curl_exec($session);
$http_status = curl_getinfo($session, CURLINFO_HTTP_CODE);
curl_close($session);

if ($http_status !== 201) {
    die("Failed to create transcription: $res");
}

$transcription = json_decode($res, true);
echo "Transcription started with ID: " . $transcription["id"] . "\n";

// 3. Poll the transcription endpoint until the status is 'completed'
while (true) {
    $session = create_curl_session("$api_base/v1/transcriptions/{$transcription["id"]}", $auth_header);
    $res = curl_exec($session);
    $http_status = curl_getinfo($session, CURLINFO_HTTP_CODE);
    curl_close($session);

    if ($http_status !== 200) {
        die("Failed to get transcription: $res");
    }

    $transcription = json_decode($res, true);

    if ($transcription["status"] === "error") {
        die("Transcription error: {$transcription["error_message"]}");
    } else if ($transcription["status"] === "completed") {
        break;
    }

    // Wait for 1 second before polling again
    sleep(1);
}

// 4. Retrieve the final transcript once transcription is completed
$session = create_curl_session("$api_base/v1/transcriptions/{$transcription["id"]}/transcript", $auth_header);
$res = curl_exec($session);
$http_status = curl_getinfo($session, CURLINFO_HTTP_CODE);
curl_close($session);

if ($http_status !== 200) {
    die("Failed to get transcript: $res");
}

$transcript = json_decode($res, true);

echo "Transcript:\n";
echo $transcript["text"] . PHP_EOL;

?>
